src/Microsoft.DocAsCode.Dfm/RendererCreator.cs

```diff
@@ -0,0 +1,182 @@
+        {
+            if (baseRenderer == null)
+            {
+                throw new ArgumentNullException(nameof(baseRenderer));
+            }
+            if (partProviders == null)
+            {
+                return baseRenderer;
+            }
+            return CreateRendererNoCheck(baseRenderer, partProviders, parameters ?? ImmutableDictionary<string, object>.Empty);
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private static ModuleBuilder CreateModule()
+        {
+            const string name = "dfm-renderer";
+            var dynamicAssembly = AppDomain.CurrentDomain.DefineDynamicAssembly(
+                new AssemblyName(name),
+                AssemblyBuilderAccess.Run);
+            return dynamicAssembly.DefineDynamicModule(name);
+        }
+
+        private static object CreateRendererNoCheck(object baseRenderer, IEnumerable<IDfmRendererPartProvider> partProviders, IReadOnlyDictionary<string, object> parameters)
+        {
+            var partGroups = from provider in partProviders
+                             from part in provider.CreateParts(parameters)
+                             select new
+                             {
+                                 part,
+                                 types = Tuple.Create(part.MarkdownRendererType, part.MarkdownTokenType, part.MarkdownContextType)
+                             } into info
+                             group info.part by info.types;
+            if (!partGroups.Any())
+            {
+                return baseRenderer;
+            }
+            foreach (var item in from g in partGroups
+                                 where !ValidateKey(g.Key)
+                                 from item in g
+                                 select item)
+            {
+                Logger.LogWarning($"Ignore invalid renderer part: {item.Name}.");
+            }
+            var type = _module.DefineType("renderer-t" + Interlocked.Increment(ref _typeCounter).ToString(), TypeAttributes.Public | TypeAttributes.Class, typeof(PlugableRendererBase));
+            var f = type.DefineField("f", typeof(IDfmRendererPart[]), FieldAttributes.Private | FieldAttributes.InitOnly);
+            DefineConstructor(type, f);
+            List<IDfmRendererPart> partList = new List<IDfmRendererPart>();
+            foreach (var g in from g in partGroups
+                              where ValidateKey(g.Key)
+                              select g)
+            {
+                DefineMethod(type, g.Key, g, f, partList);
+            }
+            return Activator.CreateInstance(type.CreateType(), baseRenderer, partList.ToArray());
+        }
+
+        private static bool ValidateKey(Tuple<Type, Type, Type> types)
+        {
+            return types.Item1 != null && types.Item2 != null && types.Item3 != null &&
+                types.Item1 != typeof(void) && types.Item2 != typeof(void) && types.Item3 != typeof(void);
```

Comment:
if it is not a correct type in business, it will never be called, but `null` and `void` will crash the creator.

---
In reply to: [110810037](https://github.com/dotnet/docfx/pull/1534#discussion_r110810037) [](ancestors = 110810037)